### DISCUSSION TITLE
Cobalt Upgrade Proposal

### DISCUSSION THREAD

**Discussion by kostko** [oasisprotocol/community-forum#18] [Mainnet]
As first announced in the [roadmap](https://github.com/oasisprotocol/community/discussions/5) and after lots of development and testing efforts we are excited to propose the first major Oasis Network upgrade named Cobalt!

To summarize, the *Oasis Core 21.1.x* release includes the following major features:

* **Light Clients and Checkpoint Sync**: In order to make bootstrapping of new network nodes much faster, the upgrade will introduce support for light clients and restoring state from checkpoints provided by other nodes in the network.

* **Random Beacon**: The random beacon is used by the consensus layer for ParaTime committee elections and is a critical component in providing security for ParaTimes with an open admission policy. The improved random beacon implementation is based on SCRAPE and provides unbiased output as long as at least one participant is honest.

* **On-Chain Governance**: The new on-chain governance service provides a simple framework for submitting governance proposals, validators voting on proposals and once an upgrade proposal passes, having a way to perform the upgrade in a controlled manner which minimizes downtime.

* **Support for Beneficiary Allowances**: Each account is able to configure beneficiaries which are allowed to withdraw tokens from it in addition to the account owner.

* **ROSE Transfers Between the Consensus Layer and ParaTimes**: The proposed upgrade introduces a mechanism where ParaTimes can emit messages as part of processing any ParaTime block. These messages can trigger operations in the consensus layer on the ParaTime's behalf. ParaTimes get their own accounts in the consensus layer which can hold ROSE and ParaTimes are able to request them be transferred to other accounts or to withdraw from other accounts when allowed via the allowances mechanism.

* **A Path Towards Self-Governing ParaTimes**: By building on the ParaTime messages mechanism, the proposed upgrade extends ParaTime governance options and enables a path towards ParaTimes that can define their own governance mechanisms.

In addition to the specified additional features we also propose the validator set size to be increased from the current 80 to 100 validators as [suggested earlier by the community](https://github.com/oasisprotocol/community/discussions/5#discussioncomment-282746).

This upgrade marks an important milestone for the Oasis Network, as it sets the foundation for unlocking the network's full capabilities. The launch of this upgrade will enable us to expand the network's toolkit with the release of the following features at a later date:

* Introducing a high-level ParaTime SDK that provides common functionalities, thereby facilitating the ParaTime developer experience.

* Introducing a JavaScript SDK that supports both the consensus layer and arbitrary ParaTimes based on the ParaTime SDK, improving the frontend developer experience.

* Introducing a token bridge between ROSE and Ethereum, unlocking liquidity across both networks.

# Next Steps

The upgrade has been extensively tested internally and on the Testnet. Barring any critical issues discovered at the last minute, we believe the upgrade is ready to be performed in April. As soon as a suitable time is proposed for the network upgrade we will be posting additional details regarding the upgrade procedure and a summary of the genesis changes.

# Behind the Name

A few months ago the community voted on a naming scheme for network upgrades. The winning theme was Rose colors, a motif that goes well with our token ROSE. The names will be alphabetical, with Amber Network being Upgrade A, Mainnet Upgrade B, and finally Cobalt as Upgrade C. Cobalt has powerful meaning as a Rose color, and fits well with the Oasis Foundation’s signature Blue Color.

*"Cobalt (or blue) roses represent mystery, the impossible, or the unattainable. Give a cobalt rose to someone and you are saying “you are extraordinarily wonderful and unique."*

**Comment by Esya** [comment]
Awesome ! Any news regarding the increase of the consensus from 80 to 100 validators ?

**Reply by kostko** [reply]
Yes, in addition to the specified additional features we also propose the validator set size to be increased from the current 80 to 100 validators as suggested earlier by the community. I have updated the proposal text to include this.

**Comment by SmartStake** [comment]
Awesome updates  @kostko 

How is April 15 at 16:00 UTC for the network upgrade?

**Reply by kostko** [reply]
After careful thought we would like to do another round of ParaTime storage migration and governance upgrade tests on the Testnet, so something like April 28th at 16:00 UTC would be a safer bet. Thoughts?

**Reply by 1xCL** [reply]
April 28th at 16:00 UTC sounds good.

**Comment by UbikCapital** [comment]
Looks good for us.

**Comment by wjdfx** [comment]
April 28th is a good date and gives some time for testing

**Comment by mrtaspinar** [comment]
> April 28th is a good date and gives some time for testing

Sounds good to what .. The Normal plan appeared to be March 31. There is no official statement that it has been postponed ..

**Comment by Afauvre** [comment]
Let's do April 28!

**Reply by mrtaspinar** [reply]
It's ok my chief. I trust you.
